drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* Set the connector, CRTC and plane properties in a single IOCTL. Here we
+	 * leave it as a blocking call, since we can't start to do page-flips before
+	 * setting everything up.
+	 */
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * Just like modeset_atomic_modeset(), in modeset_atomic_commit() we're going to
+ * atomically set some properties for our objects. The difference is that here
+ * we are setting up just the planes properties. So this is used when we want to
+ * flip a page and not during the modeset setup.
+ */
+
+static int modeset_atomic_commit(int fd, struct modeset_buf *buf, struct modeset_dev *dev)
+{
+	struct drm_object *plane = &dev->plane;
+	drmModeAtomicReq *req;
+	int ret, flags;
+
+	req = drmModeAtomicAlloc();
+
+	set_plane_property(req, plane, dev->plane_id, "FB_ID", buf->fb);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_ID", dev->crtc_id);
+	set_plane_property(req, plane, dev->plane_id, "SRC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+	set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* we want to receive the page-flip event and we want this
+	 * call to be non-blocking */
+	flags = DRM_MODE_PAGE_FLIP_EVENT | DRM_MODE_ATOMIC_NONBLOCK;
+	ret = drmModeAtomicCommit(fd, req, flags, dev);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * modeset_draw() changes just a little bit. Now we have to call the function
+ * to setup the modeset for each output.
+ * 
+ * We also reinforce that here the draw happens for the first time. In the
+ * not-atomic vsync'ed version, the KMS setup was done using drmModeSetCrtc(),
+ * which would also cause a page-flip. Using the atomic API this doesn't happen.
+ */
+
+static void modeset_draw(int fd)
+{
+	int ret;
+	fd_set fds;
+	time_t start, cur;
+	struct timeval v;
+	drmEventContext ev;
+	struct modeset_dev *iter;
+
+	/* initial modeset on all outputs; note that this is done with a blocking
+	 * call, since we can't do page-flips before setting KMS up */
+	for (iter = modeset_list; iter; iter = iter->next) {
+		modeset_atomic_modeset(fd, iter);
+	}
+
+	/* init variables */
+	srand(time(&start));
+	FD_ZERO(&fds);
+	memset(&v, 0, sizeof(v));
+	memset(&ev, 0, sizeof(ev));
+	/* Set this to only the latest version you support. Version 2
+	 * introduced the page_flip_handler, so we use that. */
+	ev.version = 2;
+	ev.page_flip_handler = modeset_page_flip_event;
```

Comment:
When you commit multiple CRTCs in a single atomic request, you would be better using version 3 `page_flip_handler2` which tells you for which CRTC the event is for. That way you can then move on to running each CRTC independently rather than committing them all at the same time limiting them all to the slowest one.

The cap `DRM_CAP_CRTC_IN_VBLANK_EVENT` must be supported for the CRTC field to be set in the flip event, see `drmGetCap()`.

But the initial modeset is important to do with all CRTCs in the same commit.

Reply:
If we decide to run with just one connector, this won't be necessary. But I prefer to do this (use multiple outputs and `page_flip_handler2`) instead, since it's good to keep a parallel with the other examples.

Reply:
Sure, go ahead. Using multiple outputs here also shows how to do it right, because doing it simply output by output like one does with the legacy API would be mostly wrong.